File: Builder.java
/*****
 * Builder
 * By separating the representation and construction of a complex class,
 * different representations can be obtained during the same construction process.
*****/


class Product {
    private String value1;
    private String value2;
    private String value3;

    public void setValue1(String newValue) { this.value1 = newValue; }
    public void setValue2(String newValue) { this.value2 = newValue; }
    public void setValue3(String newValue) { this.value3 = newValue; }

    public void method() { System.out.println(value1 + ", " + value2 + ", " + value3); }
}

abstract class Builder_ {
    protected Product product;

    public abstract void buildStep1();
    public abstract void buildStep2();
    public abstract void buildStep3();
    
    public Product getProduct() { return product; }
}

class ConcreteBuilder extends Builder_ {
    @Override
    public void buildStep1() {
        if (product == null) product = new Product();
        product.setValue1("Value1");
    }

    @Override
    public void buildStep2() {
        if (product == null) product = new Product();
        product.setValue2("Value2");
    }

    @Override
    public void buildStep3() {
        if (product == null) product = new Product();
        product.setValue3("Value3");
    }
}

class Director {
    private Builder_ builder;

    public void setBuilder(Builder_ builder) { this.builder = builder; }

    public Product construct() {
        if (builder != null) {
            builder.buildStep1();
            builder.buildStep2();
            builder.buildStep3();
        }
        
        return builder.getProduct();
    }
}


public class Builder {
    public static void main(String[] args) {
        Director d = new Director();
        d.setBuilder(new ConcreteBuilder());

        Product product = d.construct();
        product.method();
    }
}

File: Builder.cpp
/*****
 * Builder
 * By separating the representation and construction of a complex class,
 * different representations can be obtained during the same construction process.
*****/

#include <iostream>
#include <memory>

using namespace std;


class Product {
    string value1;
    string value2;
    string value3;

    public:
        void setValue1(const string& new_value) { value1 = new_value; }
        void setValue2(const string& new_value) { value2 = new_value; }
        void setValue3(const string& new_value) { value3 = new_value; }

        void method() const { cout << value1 << ", " << value2 << ", " << value3 << endl; }
};

class Builder {
    protected:
        unique_ptr<Product> product;

    public:
        Builder() = default;
        Builder(const Builder&) = delete;
        Builder& operator=(const Builder&) = delete;
        ~Builder() = default;

        virtual void buildStep1() = 0;
        virtual void buildStep2() = 0;
        virtual void buildStep3() = 0;
    
        unique_ptr<Product> getProduct() { return move(product); }
};

class ConcreteBuilder : public Builder {
    public:
        void buildStep1() override {
            if (!product) product = make_unique<Product>();
            product->setValue1("Value1");
        }

        void buildStep2() override {
            if (!product) product = make_unique<Product>();
            product->setValue2("Value2");
        }

        void buildStep3() override {
            if (!product) product = make_unique<Product>();
            product->setValue3("Value3");
        }
};

class Director {
    unique_ptr<Builder> builder;

    public:
        void setBuilder(unique_ptr<Builder> new_builder) { builder = move(new_builder); }

        unique_ptr<Product> construct() {
            if (builder) {
                builder->buildStep1();
                builder->buildStep2();
                builder->buildStep3();
            }
            
            return builder->getProduct();
        }
};


int main() {
    Director d;

    d.setBuilder(make_unique<ConcreteBuilder>());
    auto product = d.construct();
    product->method();

    return 0;
}
